### PR TITLE
MM-21343: Recognize multi-line slash commands without requiring trailing space after trigger word

### PR DESCRIPTION
**Pull request by allenlai18**
#### Summary

splitting after the first whitespace character now. 

Need some help with Unit test.

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-server/issues/13642
https://mattermost.atlassian.net/browse/MM-21343

### CODE REVIEW

File: app/command.go [hunk 1]
```diff
@@ -155,10 +155,17 @@ func (a *App) ListAllCommands(teamId string, T goi18n.TranslateFunc) ([]*model.C
 }
 
 func (a *App) ExecuteCommand(args *model.CommandArgs) (*model.CommandResponse, *model.AppError) {
-	parts := strings.Split(args.Command, " ")
-	trigger := parts[0][1:]
+	message := ""
+	trigger := ""
+	if strings.ContainsAny(args.Command, " \n\t\v\r\f") {
+		index := strings.IndexAny("args.Command", " \n\t\v\r\f")
```
Comment (lieut-data) [inline app/command.go]: The `"args.Command"` doesn't seem right -- are the unit tests failing as a result?

File: app/command.go [hunk 2]
```diff
@@ -155,10 +155,17 @@ func (a *App) ListAllCommands(teamId string, T goi18n.TranslateFunc) ([]*model.C
 }
 
 func (a *App) ExecuteCommand(args *model.CommandArgs) (*model.CommandResponse, *model.AppError) {
-	parts := strings.Split(args.Command, " ")
-	trigger := parts[0][1:]
+	message := ""
+	trigger := ""
+	if strings.ContainsAny(args.Command, " \n\t\v\r\f") {
```
Comment (lieut-data) [inline app/command.go]: It should suffice to only invoke `IndexAny` which already returns `-1` if no matches are found.

Comment (lieut-data) [reply app/command.go]: Should we consider finding a way to rely on `IsSpace` as used inside [`strings.TrimString`](https://golang.org/src/strings/strings.go?s=21924:21955#L861) to match instead?

File: app/command_test.go [hunk 3]
```diff
@@ -75,6 +75,29 @@ func TestCreateCommandPost(t *testing.T) {
...
+		Command: "/code happy path",
+	}
+
+	resp, err := th.App.ExecuteCommand(args)
+
+	assert.Nil(t, err)
+	assert.Equal(t, resp.Text, "happy path")
+
+	args = &model.CommandArgs{
+		Command: "/code\nnewline path",
```
Comment (lieut-data) [inline app/command_test.go]: I'd love to see expanded unit tests cases that test a variety of inputs, ideally using table tests to setup and define. It doesn't hurt to exhaust the potential whitespace characters as tests, as well as look for other interesting edge cases.

File: app/command.go (hunk 2 above)
Comment (allenlai18) [reply app/command.go]: suggestions taken!

File: app/command_test.go (hunk 3 above)
Comment (allenlai18) [reply app/command_test.go]: @lieut-data  its currently working when testing manually with web app.  Still working on Unit tests to get it to pass this basic case.  Still working on trying to set up delve with VS Code so i can debug.
    
Is there any good examples in the code base for using table tests?

Comment (lieut-data) [reply app/command_test.go]: For sure! Here's a good example: https://github.com/mattermost/mattermost-server/blob/ebfd332161cc8a8d62160964a5e96b82383becef/app/authentication_test.go#L16-L52

File: app/command.go [hunk 4]
```diff
@@ -155,10 +156,17 @@ func (a *App) ListAllCommands(teamId string, T goi18n.TranslateFunc) ([]*model.C
...
+	trigger := ""
+	message := ""
+	index := strings.IndexFunc(args.Command, unicode.IsSpace)
+	if index != -1 {
+		trigger = args.Command[:index]
+		message = args.Command[index+1:]
+	} else {
+		trigger = args.Command
+	}
+	trigger = trigger[1:] // remove '/'
```
Comment (lieut-data) [inline app/command.go]: Would it make sense to check for `/` explicitly? I bet this method behaves unexpectedly if a `CommandArgs` is provided without a leading slash.

Comment (allenlai18) [reply app/command.go]: yeah. I am not sure what inputs to expect. if it gets sanitized, or how this feature works end to end. or how it gets to this function ```executeCommand```   I will look into it.  

I think checking explicitly would be good. how should we treat an error? throw an exception?

Comment (lieut-data) [reply app/command.go]: I think an error probably suffices to handle this case, yes.

File: app/command.go [hunk 5]
```diff
@@ -165,8 +165,11 @@ func (a *App) ExecuteCommand(args *model.CommandArgs) (*model.CommandResponse, *
 	} else {
 		trigger = args.Command
 	}
-	trigger = trigger[1:] // remove '/'
 	trigger = strings.ToLower(trigger)
+	if trigger[0] != '/' {
+		return nil, model.NewAppError("command", "api.command.execute_command.not_found.app_error", map[string]interface{}{"Trigger": trigger}, "", http.StatusNotFound)
```
Comment (lieut-data) [inline app/command.go]: I'd suggest `http. StatusBadRequest` here, and picking an error message that explicitly indicates that the command did not include a leading slash.

File: app/command_test.go [hunk 6]
```diff
@@ -79,23 +79,31 @@ func TestExecuteCommand(t *testing.T) {
 	th := Setup(t).InitBasic()
 	defer th.TearDown()
 
-	args := &model.CommandArgs{
-		Command: "/code happy path",
+	// Testing slash command message is correctly parsed
```
Comment (lieut-data) [inline app/command_test.go]: Can we add a test for a trigger word missing the `/`?

File: app/command.go (hunk 5 above)
Comment (allenlai18) [reply app/command.go]: I would create a new entry in ```en.json``` right? something like this?

```
  {
    "id": "api.command.execute_command.format.app_error",
    "translation": "Unable to find slash character"
  }
```

Comment (lieut-data) [reply app/command.go]: Yes, that's right! Suggest text like: `Slash command trigger word missing leading slash character`

File: app/command_test.go (hunk 6 above)
Comment (allenlai18) [reply app/command_test.go]: yes

**Comment by allenlai18** [comment]
@lieut-data   sorry for the delay. this is ready for review

File: app/command.go [hunk 7]
```diff
@@ -155,10 +156,20 @@ func (a *App) ListAllCommands(teamId string, T goi18n.TranslateFunc) ([]*model.C
...
+	index := strings.IndexFunc(args.Command, unicode.IsSpace)
+	if index != -1 {
+		trigger = args.Command[:index]
+		message = args.Command[index+1:]
+	} else {
+		trigger = args.Command
+	}
 	trigger = strings.ToLower(trigger)
-	message := strings.Join(parts[1:], " ")
+	if trigger[0] != '/' {
```
Comment (lieut-data) [inline app/command.go]: Let's use `strings.HasPrefix` to make the code resilient to an empty `args.Command` which might otherwise trigger an array out of bounds error here.
```suggestion
	if !strings.HasPrefix(trigger, "/") {
```

File: app/command.go [hunk 8]
```diff
@@ -155,10 +156,20 @@ func (a *App) ListAllCommands(teamId string, T goi18n.TranslateFunc) ([]*model.C
...
+		return nil, model.NewAppError("command", "api.command.execute_command.format.app_error", map[string]interface{}{"Trigger": trigger}, "", http.StatusBadRequest)
+	}
+	trigger = trigger[1:] // remove '/'
```
Comment (lieut-data) [inline app/command.go]:
```suggestion
	trigger = strings.TrimPrefix(trigger, "/")
```

File: app/command_test.go [hunk 9]
```diff
@@ -75,6 +75,45 @@ func TestCreateCommandPost(t *testing.T) {
 	}
 }
 
+func TestExecuteCommand(t *testing.T) {
+	th := Setup(t).InitBasic()
+	defer th.TearDown()
+
+	// Testing slash command message is correctly parsed
+	TestCases := []string{
```
Comment (lieut-data) [inline app/command_test.go]: A few code formatting suggestions. I think the tests would be easier to manage as a map, since the input and expected output are closer together.
```go
testCases := map[string]string {
    "/code happy path": "    happy path",
    // ...
}
```

File: app/command_test.go [hunk 10]
```diff
@@ -75,6 +75,45 @@ func TestCreateCommandPost(t *testing.T) {
...
+	for index, TestCase := range TestCases {
+		args := &model.CommandArgs{
+			Command: TestCase,
+			T:       func(s string, args ...interface{}) string { return s },
+		}
+		resp, _ := th.App.ExecuteCommand(args)
+
+		assert.Equal(t, resp.Text, results[index])
+	}
+	argsMissingSlashCharacter := &model.CommandArgs{
```
Comment (lieut-data) [inline app/command_test.go]: suggest wrapping this in a subtest:
```go
t.Run("missing slash character", func(t *testing.T) {
    	argsMissingSlashCharacter := &model.CommandArgs{
                // ...
        }
        // ...
})
```

and then the above block in a subtest named `valid commands` or some such.

Comment (lieut-data) [reply app/command_test.go]: Given my `strings.HasPrefix` comment above, can we also add one more test that asserts the behaviour of an empty `args.Command` string?

File: app/command_test.go (hunk 9 above)
Comment (allenlai18) [reply app/command_test.go]: you are absolutely right

File: app/command_test.go (hunk 10 above)
Comment (allenlai18) [reply app/command_test.go]: absolutely

File: app/command_test.go [hunk 11]
```diff
@@ -74,6 +74,38 @@ func TestCreateCommandPost(t *testing.T) {
 	require.Equal(t, err.Id, "api.context.invalid_param.app_error")
 }
 
+func TestExecuteCommand(t *testing.T) {
+	th := Setup(t).InitBasic()
+	defer th.TearDown()
+
+	// Testing slash command message is correctly parsed
+	TestCases := map[string]string {
```
Comment (golangcibot) [inline app/command_test.go]: File is not `gofmt`-ed with `-s` (from `gofmt`)

**Comment by allenlai18** [comment]
@lieut-data  All your suggestions are absolutely correct. Thanks for showing me how to improve little things! Please review again when you get a chance

File: app/command_test.go [hunk 12]
```diff
@@ -74,6 +74,53 @@ func TestCreateCommandPost(t *testing.T) {
 	require.Equal(t, err.Id, "api.context.invalid_param.app_error")
 }
 
+func TestExecuteCommand(t *testing.T) {
+	th := Setup(t).InitBasic()
+	defer th.TearDown()
+
+	t.Run("valid tests with different whitespace characters", func(t *testing.T) {
+		TestCases := map[string]string {
```
Comment (golangcibot) [inline app/command_test.go]: File is not `gofmt`-ed with `-s` (from `gofmt`)

**Comment by allenlai18** [comment]
@hanzei   I think we can remove dev review label and merge?

File: app/command.go [hunk 13]
```diff
@@ -9,6 +9,7 @@ import (
 	"net/http"
 	"net/url"
 	"strings"
+	"unicode"
```
Comment (golangcibot) [inline app/command.go]: File is not `gofmt`-ed with `-s` (from `gofmt`)
```suggestion

```

**Comment by hanzei** [comment]
@lieut-data I'm wondering if this change should go into 5.22. Thoughts?

**Comment by lieut-data** [comment]
I think that's fine, given this is not a major feature, and we're well before code complete.
